Initial KE100 support.

pwoerndle opened the pull request:
Reading data works. Setting temperature and frost protection returns empty result from the device, which creates an EmptyResult runtime error.

pwoerndle commented:
I did some more digging and the EmptyResult error is thrown in control_child method in `tapo/src/api_client.rs` on the last statement.

```rust
    pub(crate) async fn control_child<R>(
        &self,
        device_id: String,
        child_request: TapoRequest,
    ) -> Result<R, Error>
    where
        R: fmt::Debug + DeserializeOwned + TapoResponseExt,
    {
        debug!("Control child...");
        let params = MultipleRequestParams::new(vec![child_request]);
        let request = TapoRequest::MultipleRequest(Box::new(TapoParams::new(params)));

        let params = ControlChildParams::new(device_id, request);
        let request = TapoRequest::ControlChild(Box::new(TapoParams::new(params)));

        let responses = self
            .protocol
            .execute_request::<ControlChildResult<TapoMultipleResponse<R>>>(request, true)
            .await?
            .ok_or_else(|| Error::Tapo(TapoResponseError::EmptyResult))?
            .response_data
            .result
            .responses;

        let response = responses
            .into_iter()
            .next()
            .ok_or_else(|| Error::Tapo(TapoResponseError::EmptyResult))?;

        validate_response(&response)?;

        response
           .result
           .ok_or_else(|| Error::Tapo(TapoResponseError::EmptyResult))
    }
```

Since the control_child method is used for read and write operations, I guess we need to check the result to be empty at some point, but would it make sense to do the check, when we interpret the result?

mihai-dinculescu commented:
Looking at the tapo app source code, I see a few fields that might be useful:
- `temp_unit` string (it should probably be converted to an enum)
- `min_control_temp` integer
- `max_control_temp` integer
- `min_temp` integer (it ends up getting used by frost protection)

You're correct. `ApiClient::control_child` should either be updated to handle empty responses or a copy of the function that can should be created.

pwoerndle commented:
This what I get when executing `device.set_temperature(19).await?;`:

```
 2023-10-20T06:07:48.307Z INFO  tapo_ke100                                > EXECUTING COMMAND WITH EMPTY RESULT
 2023-10-20T06:07:48.307Z DEBUG tapo::api::api_client                     > Control child...
 2023-10-20T06:07:48.307Z DEBUG tapo::api::protocol::passthrough_protocol > Request to passthrough: {"method":"control_child","params":{"device_id":"REDACTED","requestData":{"method":"multipleRequest","params":{"requests":[{"method":"set_device_info","params":{"device_on":true,"target_temp":19}}]}}}}
 2023-10-20T06:07:50.506Z DEBUG tapo::api::protocol::passthrough_protocol > Device responded with: TapoResponse { error_code: 0, result: Some(TapoResult { response: "kSb5Ej1g6e5y/FqhXbR8zs2pbesSXEoFnpur25O4SvYOAZIpvJ6bXrBACLOoaxeYHShR60TbXG1bad+r/jBY1vjKaAvEN3TYXqkxTTHp051EIBT88l1z1rV538LWPgGcLy+hD1rUdDX0wMO/p7HvDwMPIv6wxhnG/NwGxaTxt8g=" }) }
 2023-10-20T06:07:50.506Z DEBUG tapo::api::protocol::passthrough_protocol > Device inner response decrypted: {"result":{"responseData":{"result":{"responses":[{"method":"set_device_info","error_code":0}]}}},"error_code":0}
 2023-10-20T06:07:50.507Z DEBUG tapo::api::protocol::passthrough_protocol > Device inner response: TapoResponse { error_code: 0, result: Some(ControlChildResult { response_data: TapoMultipleResponse { result: TapoMultipleResult { responses: [TapoResponse { error_code: 0, result: None }] } } }) }
 2023-10-20T06:07:50.507Z DEBUG tapo::api::protocol::passthrough_protocol > validate_response passed
 2023-10-20T06:07:50.507Z DEBUG tapo::api::protocol::passthrough_protocol > assign result passed
```

What do you think about splitting `ApiClient::control_child` into `ApiClient::read_child` and `ApiClient::write_child` to handle the `None` response for the write operation? I'm struggling a bit to make `ApiClient::control_child` handle both cases.

mihai-dinculescu commented:
This should get the job done.

api_client

```rust
    pub(crate) async fn control_child<R>(
        &self,
        device_id: String,
        child_request: TapoRequest,
    ) -> Result<Option<R>, Error>
    where
        R: fmt::Debug + DeserializeOwned + TapoResponseExt,
    {
        debug!("Control child...");
        let params = MultipleRequestParams::new(vec![child_request]);
        let request = TapoRequest::MultipleRequest(Box::new(TapoParams::new(params)));

        let params = ControlChildParams::new(device_id, request);
        let request = TapoRequest::ControlChild(Box::new(TapoParams::new(params)));

        let responses = self
            .protocol
            .execute_request::<ControlChildResult<TapoMultipleResponse<R>>>(request, true)
            .await?
            .ok_or_else(|| Error::Tapo(TapoResponseError::EmptyResult))?
            .response_data
            .result
            .responses;

        let response = responses
            .into_iter()
            .next()
            .ok_or_else(|| Error::Tapo(TapoResponseError::EmptyResult))?;

        validate_response(&response)?;

        Ok(response.result)
    }
```

hub_handler

```rust
    pub(crate) async fn control_child<R>(
        &self,
        device_id: String,
        request_data: TapoRequest,
    ) -> Result<Option<R>, Error>
    where
        R: fmt::Debug + DeserializeOwned + TapoResponseExt,
    {
        self.client.control_child(device_id, request_data).await
    }
```

s200b_handler

```rust
    pub async fn get_device_info(&self) -> Result<S200BResult, Error> {
        let request = TapoRequest::GetDeviceInfo(TapoParams::new(EmptyParams));

        self.hub_handler
            .control_child(self.device_id.clone(), request)
            .await?
            .ok_or_else(|| Error::Tapo(TapoResponseError::EmptyResult))
    }
```

pwoerndle commented:
Thanks for the code snippets. I managed to resolve the issue with the `None` result in 8716347a4efdd0f6021316116e89eacc6218fde1. Need to work a bit more on this to support temperatures in float and add the parameters you suggested above.

pwoerndle commented:
Meanwhile I added support for min/max_control temp, child_protection and temp_offset.
I added a `min_temp` parameter, but although the API doesn't respond with an error when setting a value the value on the device doesn't change (checked in the app). What is strange, is that neither device_info nor component_list seem to expose the min_temp parameter.
I'm thinking of adding a function to query the current temp directly for convenience as this will most likely be a common use case, apart from setting the temperature.

Let me know if you have any other thoughts on what could be added.

mihai-dinculescu commented:
You're correct.
There's a separate endpoint called `set_frost_protection` that requires `min_temp` and `temp_unit` as parameters. It's up to you if you want to include it in this pull request. Alternatively, we could save it for a future update to keep this one more manageable.
By the way, there's also a `get_frost_protection` endpoint that I suspect will return the `min_temp`.

However, It would be useful to include `temp_unit`.

Regarding your question, there's a separate endpoint called `get_temperature_records` that should fulfill the task. You can implement it in a similar way to `get_trigger_logs`.

pwoerndle commented:
I've tried the changes you proposed but seem to be getting a compilation error in `tapo/src/api/child_devices/ke100_handler.rs` now:

```
error[E0698]: type inside `async fn` body must be known in this context
  --> tapo/src/api/child_devices/ke100_handler.rs:63:14
   |
63 |             .control_child(self.device_id.clone(), request)
   |              ^^^^^^^^^^^^^ cannot infer type for type parameter `R` declared on the method `control_child`
   |
note: the type is part of the `async fn` body because of this `await`
```

I suggest we try to get this one fixed and then close this PR. I'll keep working on the other stuff and will create another PR once ready.

Thanks for the help with getting the Rust code right. As you may have recognized I'm learning by doing here.

pwoerndle commented:
Never mind, I fixed it. If I understood the Rust documentation right, I needed to bind control_child: `.control_child::<KE100Result>`.

```rust
    pub async fn set_temperature(&self, target_temperature: u8) -> Result<(), Error> {

        let control_range = self.get_control_range().await?;

        if target_temperature < control_range[0] || target_temperature > control_range[1] {
            return Err(Error::Validation {
                field: "target_temperature".to_string(),
                message: format!("Target temperature must be between {} (min_control_temp) and {} (max_control_temp)", control_range[0], control_range[1]),
            });
        }

        let json = serde_json::to_value(TrvSetDeviceInfoParams::new().target_temp(target_temperature)?)?;
        let request = TapoRequest::SetDeviceInfo(Box::new(TapoParams::new(json)));

        self.hub_handler
            .control_child::<KE100Result>(self.device_id.clone(), request)
            .await?;

        Ok(())
    }
```

I assume I can use the same pattern for all the functions which don't return a result?

mihai-dinculescu commented:
> I assume I can use the same pattern for all the functions which don't return a result?

Exactly, yeah. However, it would be better to use `.control_child::<serde_json::Value>` instead of `.control_child::<KE100Result>`.

Please give me a shout when you're done with everything you want to do, and I'll do a review.
If you don't have the time to address all the little things, you can [permit me](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to the PR.

pwoerndle commented:
Just pushed the last updates. I think the PR is ready to go for review.

mihai-dinculescu commented on tapo/src/requests/set_device_info/trv.rs:
```diff
@@ -0,0 +1,79 @@
+use crate::error::Error;
+
+#[derive(Debug, Default, Serialize)]
+pub(crate) struct TrvSetDeviceInfoParams {
```
The fields inside the struct shouldn't need to be `pub`.

pwoerndle replied:
Implemented in 0dead6c.

mihai-dinculescu reviewed (commented):
Very well done. You have addressed all the important aspects, which is impressive considering you are new to Rust.

I am aware that I left a few comments, but they are all minor and mostly related to naming and unnecessary elements that can be removed.

pwoerndle commented on tapo/src/api/child_devices/ke100_handler.rs:
```diff
@@ -0,0 +1,176 @@
+        Ok(())
+    }
+
+    /// Returns *min_control_temp* and *max_control_temp* as Vec<u8>.
+    async fn get_control_range(&self) -> Result<Vec<u8>, Error> {
+            let request = TapoRequest::GetDeviceInfo(TapoParams::new(EmptyParams));
+    
+            self.hub_handler
+                .control_child::<KE100Result>(self.device_id.clone(), request)
+                .await?
+                .ok_or_else(|| Error::Tapo(TapoResponseError::EmptyResult))
+                .map(|result| vec![result.min_control_temp,result.max_control_temp])
+        }
```
I need `get_control_range()` in `set_temperature()` to check if `target_temperature` is in the allowed range and present a proper error message with the correct range. Otherwise the device just sends an error code, which doesn't give an indication that `target_temperature` is out of the allowed range. I thus suggest to keep the function.

mihai-dinculescu replied:
Yes, but it would be simpler to call the existing `get_device_info` function inside `set_temperature()` and use `device_info. min_control_temp` and `device_info. max_control_temp` in the `if` clause directly.

pwoerndle replied:
Yes, makes total sense to use device_info directly. I'll update the code.

pwoerndle commented on tapo/src/api/child_devices/ke100_handler.rs:
```diff
@@ -0,0 +1,176 @@
+        Ok(())
+    }
+
+    /// Sets the *min temperature*, which is applied in frost protection mode.
+    ///
+    /// # Arguments
+    ///
+    /// * `min_temperature` - between 5 and 15
+    pub async fn set_min_temperature(&self, min_temperature: u8) -> Result<(), Error> {
+        let json = serde_json::to_value(TrvSetDeviceInfoParams::new().min_temp(min_temperature)?)?;
+        let request = TapoRequest::SetDeviceInfo(Box::new(TapoParams::new(json)));
+
+        self.hub_handler
+            .control_child::<serde_json::Value>(self.device_id.clone(), request)
+            .await?;
+
+        Ok(())
+    }
```
Yes, this can be removed for now.

pwoerndle commented on tapo/src/responses/child_device_list_result/ke100_result.rs:
```diff
@@ -0,0 +1,61 @@
+    pub r#type: String,
+    #[serde(rename = "temp_unit")]
+    pub temperature_unit: TemperatureUnit,
+    pub current_temp: f32,
+    pub target_temp: f32,
+    pub min_control_temp: u8,
+    pub max_control_temp: u8,
+    pub frost_protection_on: bool,
+    pub location: String,
+    pub temp_offset: i8,
```
I aligned the names with the names in the API. Renaming them would make it inconsistent with the names in the device API and thus in the json output. I don't have a strict preference though.

mihai-dinculescu replied:
In that case `temperature_unit` should be changed to `temp_unit`.

mihai-dinculescu replied:
I'm suggesting this because I've renamed all the `temp` instances to `temperature` for the T310 and T315 devices, and it would be nice to stay consistent.

pwoerndle replied:
Ok, the hint with the t31x devices helped me to understand how to do the renaming for the API calls with `serde`. Implemented the change in 0dead6c.

pwoerndle commented on tapo/src/api/child_devices/ke100_handler.rs:
```diff
@@ -0,0 +1,176 @@
+        Ok(())
+    }
+
+    /// Sets the *temperature offset*.
```
In case you have the device (and thus the radiator) close to a window, the temperature measured at the device itself may not reflect the actual room temperature. Thus you can set an offset to align with your room temperature.

mihai-dinculescu replied:
Makes sense. Thank you!

pwoerndle commented on tapo/src/api/child_devices/t31x_handler.rs:
```diff
@@ -37,8 +38,9 @@ impl<'h> T31XHandler<'h> {
+            .ok_or_else(|| Error::Tapo(TapoResponseError::EmptyResult));
 
-        Ok(result.try_into()?)
+        Ok(result.unwrap().try_into()?)
```
Implemented in 0dead6c.

pwoerndle commented on tapo/src/responses/child_device_list_result.rs:
```diff
@@ -58,6 +60,8 @@ pub enum ChildDeviceResult {
     T310(Box<T31XResult>),
     /// T315 temperature & humidity sensor.
     T315(Box<T31XResult>),
+    /// KE100 TRV.
```
Implemented in 0dead6c.

pwoerndle commented on tapo/src/responses/child_device_list_result/ke100_result.rs:
```diff
@@ -0,0 +1,61 @@
+pub enum TemperatureUnit {
+    Celsius,
+    Fahrenheit,
+}
```
Implemented in 0dead6c.

pwoerndle commented on tapo/examples/tapo_ke100.rs:
```diff
@@ -0,0 +1,44 @@
+    let ip_address = env::var("IP_ADDRESS")?;
+    // ID of the KE100 device. Can be obtained from executing `get_child_device_component_list_json()`` on the hub device.
+    let device_id = env::var("DEVICE_ID")?;
+    let target_temp: u8 = env::var("TEMPERATURE")?.parse().unwrap();
```
Implemented in 0dead6c.

pwoerndle commented on tapo/examples/tapo_ke100.rs:
```diff
@@ -0,0 +1,44 @@
+    let tapo_username = env::var("TAPO_USERNAME")?;
+    let tapo_password = env::var("TAPO_PASSWORD")?;
+    let ip_address = env::var("IP_ADDRESS")?;
+    // ID of the KE100 device. Can be obtained from executing `get_child_device_component_list_json()`` on the hub device.
```
Implemented in 0dead6c.

pwoerndle commented on tapo/src/api/child_devices/ke100_handler.rs:
```diff
@@ -0,0 +1,176 @@
+    /// # Arguments
+    ///
+    /// * `temp_offset` - between 5 and 30
+    pub async fn set_temp_offset(&self, temp_offset: i8) -> Result<(), Error> {
```
Implemented in 0dead6c.

pwoerndle commented on tapo/src/requests/set_device_info/trv.rs:
```diff
@@ -0,0 +1,79 @@
+use crate::error::Error;
+
+#[derive(Debug, Default, Serialize)]
+pub(crate) struct TrvSetDeviceInfoParams {
```
Implemented in 0dead6c.

mihai-dinculescu commented on tapo/src/requests/set_device_info/trv.rs:
```diff
@@ -0,0 +1,75 @@
+    
+    pub fn validate(self) -> Result<Self, Error> {
+        if let Some(temp_offset) = self.temperature_offset {
+            if temp_offset < -10 || temp_offset> 10 {
+                return Err(Error::Validation {
+                    field: "temp_offset".to_string(),
+                    message: "must be between -10 and 10".to_string(),
+                    
+                });
+            }
```
The comment of the `set_temperature_offset` says that the temperature offset must be between 5 and 30. But this rule checks that it's between -10 and 10. Which one is accurate?

Could this differ based on the temperature unit?

pwoerndle replied:
Good catch. The comment is a copy paste issue. I'll fix that. The -10/10 I got from what the app let's me select for the temperature offset. It's fixed to Celsius in my app, but I'd assume that the Fahrenheit values may differ. For some reason I can't get my app to let me select Fahrenheit as the temperature unit.

mihai-dinculescu replied:
The temperature settings for my other devices are located in the Device Settings.
![image](https://github.com/mihai-dinculescu/tapo/assets/9944681/e1701c97-da9d-481b-aa8b-4627695bbaa8)

pwoerndle replied:
I don’t get this option for the KE100 device. Very strange. Checked with the android and iOS app and the documentation and there seems to be no option to change that parameter. The temperature unit is not mentioned for this device.

mihai-dinculescu replied:
Looking at [this video](https://www.youtube.com/watch?v=xStn7GV0PSk), it looks like the unit is shown, but maybe not configurable?

![image](https://github.com/mihai-dinculescu/tapo/assets/9944681/cfd340d0-bd15-4059-958a-33c6d01511ae)
![image](https://github.com/mihai-dinculescu/tapo/assets/9944681/7b27e43b-d19d-4de9-b502-cc75885bca2b)

mihai-dinculescu replied:
Can you please add `temp_unit` to `TrvSetDeviceInfoParams` and check if setting it does anything?

pwoerndle replied:
Setting `temp_unit` to "celsius" works. Setting it to "fahrenheit" results in a Error: Tapo(Unknown(-1001)).

```
DEBUG tapo::api::api_client                     > Control child...
DEBUG tapo::api::protocol::passthrough_protocol > Request to passthrough: {"method":"control_child","params":{"device_id":"[REDACTED]","requestData":{"method":"multipleRequest","params":{"requests":[{"method":"set_device_info","params":{"temp_unit":"celsius"}}]}}}}
DEBUG tapo::api::protocol::passthrough_protocol > Device responded with: TapoResponse { error_code: 0, result: Some(TapoResult { response: "[REDACTED]" }) }
DEBUG tapo::api::protocol::passthrough_protocol > Device inner response decrypted: {"result":{"responseData":{"result":{"responses":[{"method":"set_device_info","error_code":0}]}}},"error_code":0}
DEBUG tapo::api::protocol::passthrough_protocol > Device inner response: TapoResponse { error_code: 0, result: Some(ControlChildResult { response_data: TapoMultipleResponse { result: TapoMultipleResult { responses: [TapoResponse { error_code: 0, result: None }] } } }) }

DEBUG tapo::api::api_client                     > Control child...
DEBUG tapo::api::protocol::passthrough_protocol > Request to passthrough: {"method":"control_child","params":{"device_id":"[REDACTED]","requestData":{"method":"multipleRequest","params":{"requests":[{"method":"set_device_info","params":{"temp_unit":"fahrenheit"}}]}}}}
DEBUG tapo::api::protocol::passthrough_protocol > Device responded with: TapoResponse { error_code: 0, result: Some(TapoResult { response: "[REDACTED]" }) }
DEBUG tapo::api::protocol::passthrough_protocol > Device inner response decrypted: {"result":{"responseData":{"result":{"responses":[{"method":"set_device_info","error_code":-1001}]}}},"error_code":0}
DEBUG tapo::api::protocol::passthrough_protocol > Device inner response: TapoResponse { error_code: 0, result: Some(ControlChildResult { response_data: TapoMultipleResponse { result: TapoMultipleResult { responses: [TapoResponse { error_code: -1001, result: None }] } } }) }
Error: Tapo(Unknown(-1001))
```

So my assumption is that it shows the unit in the app and has the field in the API, but `temp_unit` is not configurable for the KE100. I suggest I'll update the documentation in the code regarding this. Maybe worthwhile to also put that in the main readme.

mihai-dinculescu replied:
Thank you for taking the time to look into this and getting to the bottom.
Thermostat temperature is an important topic, and I agree that the library must be as obvious as possible.

How about adding a `temperature_unit` parameter to all the functions that change temperature (`set_target_temperature`, `set_min_control_temperature` and `set_max_control_temperature.`). It can be a separate enum (like it used to be) with just one value, Celsius.

Forcing users to specify the unit, even if there's only one option, will make it very obvious.

pwoerndle replied:
Sounds reasonable. I'll update the respective functions.

pwoerndle replied:
a9ebad1040a2861c931b93b6e11a153d28c49c75 now implements a specific `temperature_unit` parameter for the functions setting a temperature.

mihai-dinculescu commented on tapo/src/api/child_devices/ke100_handler.rs:
```diff
@@ -0,0 +1,149 @@
+    /// # Arguments
+    ///
+    /// * `child_protection_on` - true/false
+    pub async fn set_child_protection(&self, child_protection_on: bool) -> Result<(), Error> {
+        let json = serde_json::to_value(TrvSetDeviceInfoParams::new().child_protection(child_protection_on)?)?;
+        let request = TapoRequest::SetDeviceInfo(Box::new(TapoParams::new(json)));
+    
+        self.hub_handler
+            .control_child::<serde_json::Value>(self.device_id.clone(), request)
+            .await?;
+
+        Ok(())
+    }
```
Does this work? Looking at the decompiled Android app, it shouldn't.

pwoerndle replied:
Yes, this works. In the iOS app I can also change the parameter.

pwoerndle commented:
I found one more thing I may need your help with. It's about decoding the base64 encoded nickname. The following fragment in `example/tapo_ke100.rs` returns the base64 encoded nickname:

```rust
let hub = ApiClient::new(tapo_username, tapo_password)?
    .h100(ip_address)
    .await?;

// Get a handler for the child device
let device = hub.ke100(device_id);

// Get the device info of the child device
let device_info = device.get_device_info().await?;
info!("Device info: {device_info:?}");
```

But when I use the detour via `get_child_device_list()` as in the following fragment, I get the decoded nickname in `device.nickname`

```rust
// Adapted code from examples/tapo_h100.rs

let hub = ApiClient::new(tapo_username, tapo_password)?
    .h100(ip_address)
    .await?;
let child_device_list = hub.get_child_device_list().await?;

for child in child_device_list {
    match child {
        ChildDeviceResult::KE100(device) => {
            info!(
                "Found KE100 child device with nickname: {}, id: {}, current temperature: {} {:?} and target temperature: {} {:?}.",
                device.nickname,
                device.device_id,
                device.current_temperature,
                device.temperature_unit,
                device.target_temperature,
                device.temperature_unit,
            );
        }
```

Do you see a similar behavior, when you test other child devices such as s200b or t300? Is this intended behavior or should we do something about this?

mihai-dinculescu commented:
Ah, that is a defect. Nice find!
I've fixed it in https://github.com/mihai-dinculescu/tapo/commit/fa73e407e578dc748a4a4bca9f6528f4a18649e4.

pwoerndle commented:
Merged your fixes for the nickname decoding into this PR. The decoding seems to work also with the feature I added to supported empty results for the `control_child` function.

mihai-dinculescu commented:
I've committed some minor tweaks for language and lints.
I think this is in a good shape to ship 🚀

mihai-dinculescu commented:
Released in [v0.7.6](https://github.com/mihai-dinculescu/tapo/releases/tag/v0.7.6).
Thank you for this amazing effort!
